c/tskit/trees.c

```diff
@@ -641,9 +641,10 @@ increment_nd_array_value(double *array, tsk_size_t n, const tsk_size_t *shape,
 /* TODO flatten the reference sets input here and follow the same pattern used
  * in diversity, divergence, etc. */
 int TSK_WARN_UNUSED
-tsk_treeseq_genealogical_nearest_neighbours(tsk_treeseq_t *self, tsk_id_t *focal,
-    size_t num_focal, tsk_id_t **reference_sets, size_t *reference_set_size,
-    size_t num_reference_sets, tsk_flags_t TSK_UNUSED(options), double *ret_array)
+tsk_treeseq_genealogical_nearest_neighbours(const tsk_treeseq_t *self,
+    const tsk_id_t *focal, size_t num_focal, tsk_id_t *const *reference_sets,
```

Comment:
I see what you mean about the trickiness of this one @castedo. In this case, I think it's probably simplest to skip the const-ing of the ``reference_sets`` here and in ``mean_descendents``, because we want to change it anyway to use a flattened input of ``sample_sets`` like the ``diversity``, divergence`` etc. I've opened #948

Reply:
:+1: on #948, I bet that will be less confusing for most readers.

Until then, I see two sub-optimal options
1) use `*const` which might confuse some readers who are not familiar with `*const` or
2) don't use `*const` which will confuse readers who are familiar with it because the API is declaring that the function WILL change that parameter (all the other read-only params use const, it would stand out that this one isn't marked read-only)

I'm inclined to do 2) following the principle of using the "vocabulary/words" precisely and make those unfamiliar with that "vocabulary/word" look them up.

Reply:
(2) sounds good. Hopefully these functions will be updated to use the correct API soon anyway.